### DISCUSSION TITLE
Cookie Component

### DISCUSSION THREAD

**Discussion by tomtomkratos** [sqlpage/SQLPage#684] [General]
Hi, there,
I've set up a cookie component to manage the selection of a ‘dark’ theme and its deselection. It works great.
But only on the computer where the sqlpage server is running. When I connect to the server with a smartphone for example, I can't change the theme... 
I don't know how to do it 
tom

**Comment by YunusMechery** [comment]
Cookies are browser specific. 
When you **try in a incognito window**, in the computer where sqlpage is running, what you get?

**Comment by tomtomkratos** [comment]
In a incognito window, in the computer where sqlpage is running, the ‘dark’ theme managed by the **sqlpage.cookie** function works perfectly.

**Reply by lovasoa** [reply]
Your mobile browser is probably configured to refuse cookies...

**Comment by tomtomkratos** [comment]
I have tried several things : 
- the computer where sqlpage is running : OK
- with another computer connected to the server: NOPE
- on a connected mobile phone: NOPE
I paid attention to cookie activation. 
Maybe I wrote the sql code wrong. However, I used the example [(https://github.com/sqlpage/SQLPage/tree/main/examples/light-dark-toggle)](https://github.com/sqlpage/SQLPage/tree/main/examples/light-dark-toggle)
I'm not an expert on this cookie/server subject :smile:.

**Reply by lovasoa** [reply]
## Secure cookies

You should also be careful about security contexts.

As explained in the documentation, cookies are *secure* by default, which means they are only transmitted either to localhost, or to the website **exclusively over https**.

If you are connecting the smartphone to the computer where sqlpage is running over a local network, by entering an URL like `http://192.168.0.1` in the address bar of your mobile device, then you are connecting over **http**, not **https**. You cannot transmit secure cookies over **http**. You have to either deploy your website over **https**, or make the cookie insecure (which should not be a problem if it stores only a *dark/light* state toggle, and not something sensitive like a session token). 

You can make your cookies insecure if you want, allowing them to be transferred over http connections, by setting `FALSE as secure` in the cookie component. 

> ```sql
> SELECT 'cookie' as component, 'username' as name, 'John Doe' as value,
>       FALSE AS secure; -- You can remove this if the site is served over HTTPS.
>  ```

https://sql.datapage.app/component.sql?component=cookie#component

https://en.wikipedia.org/wiki/Secure_cookie

## Testing

To check that you can store cookies you can build a simple set of two pages:

### `write.sql`

```sql
SELECT 'cookie' as component,
        'username' as name,
        :username as value
where :username is not null;

select 'form' as component;
select 'username' as name;
```

### `read.sql`

```sql
SELECT 'text' as component,
        'Hello, ' || sqlpage.cookie('username') || '!' as contents;
```

**Comment by tomtomkratos** [comment]
***Many thanks!!!*** 👍 
I had totally forgotten to use this top-level parameter.
```sql
FALSE as secure
```
Now it works perfectly !!!
